### PR TITLE
i-cbmc: add support for verification of nested interrupts

### PR DESCRIPTION
**Pull request by lihaol**
This series of patches adds nested-interrupts verification to CBMC's concurrency features. For more information, please see http://www.cprover.org/interrupts/

Usage: cbmc --mm irq main.c

### CODE REVIEW

File: src/goto-symex/memory_model_interrupt.cpp [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+\*******************************************************************/
+
+void memory_model_interruptt::read_from(symex_target_equationt &equation)
+{
+  memory_model_sct::read_from(equation);
+
+  for(choice_symbolst::const_iterator
+      c_it=choice_symbols.begin();
+      c_it!=choice_symbols.end();
+      c_it++)
```
Comment (tautschnig) [inline src/goto-symex/memory_model_interrupt.cpp]: Should use a ranged for

File: src/goto-symex/symex_function_call.cpp [hunk 2]
```diff
@@ -7,6 +7,7 @@ Author: Daniel Kroening, kroening@kroening.com
 \*******************************************************************/
 
 #include <cassert>
+#include <iostream>
```
Comment (tautschnig) [inline src/goto-symex/symex_function_call.cpp]: Only with #ifdef DEBUG

Comment (lihaol) [inline src/goto-symex/symex_function_call.cpp]: Thanks for the reviews. All done. Please see my new commits.

**Comment by kroening** [comment]
And now needs another rebase!
Furthermore, I am wondering whether this should best live in a branch.

**Comment by tautschnig** [comment]
@kroening Why would you prefer a branch?

**Comment by lihaol** [comment]
I have done a rebase.

File: src/goto-symex/symex_function_call.cpp [hunk 3]
```diff
@@ -9,6 +9,9 @@ Author: Daniel Kroening, kroening@kroening.com
 #include <iostream>
 #include <sstream>
 #include <cassert>
+#ifdef DEBUG
+#include <iostream>
+#endif
```
Comment (tautschnig) [inline src/goto-symex/symex_function_call.cpp]: There already is an unconditional `#include <iostream>` above - I suppose that should go away?

File: src/goto-symex/memory_model_interrupt.h [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+#ifndef CPROVER_MEMORY_MODEL_INTERRUPT_H
+#define CPROVER_MEMORY_MODEL_INTERRUPT_H
+
+#include "memory_model_sc.h"
+
+class memory_model_interruptt:public memory_model_sct
+{
+public:
+  inline explicit memory_model_interruptt(const namespacet &_ns):
```
Comment (tautschnig) [inline src/goto-symex/memory_model_interrupt.h]: No need for `inline` in the class definition.

File: src/goto-symex/memory_model_interrupt.h [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  void from_read(symex_target_equationt &equation);
+  void write_serialization_external(symex_target_equationt &equation);
+
+private:
+  per_thread_mapt per_thread_map;
+  exprt last(const event_it &from, const event_it &to);
+};
+
+#endif
+
```
Comment (tautschnig) [inline src/goto-symex/memory_model_interrupt.h]: No blank line at end-of-file, please.

File: src/goto-symex/memory_model_interrupt.h [hunk 6]
```diff
@@ -0,0 +1,35 @@
+/*******************************************************************\
+
+Module: Memory models for partial order concurrency
+
+Author: Lihao Liang, lihao.liang@cs.ox.ac.uk
+
+\*******************************************************************/
+
+#ifndef CPROVER_MEMORY_MODEL_INTERRUPT_H
+#define CPROVER_MEMORY_MODEL_INTERRUPT_H
```
Comment (tautschnig) [inline src/goto-symex/memory_model_interrupt.h]: That should be "CPROVER_GOTO_SYMEX_MEMORY_MODEL_INTERRUPT_H"

File: src/cbmc/bmc.cpp [hunk 7]
```diff
@@ -422,10 +423,12 @@ safety_checkert::resultt bmct::run(
     memory_model=std::unique_ptr<memory_model_baset>(new memory_model_tsot(ns));
   else if(mm=="pso")
     memory_model=std::unique_ptr<memory_model_baset>(new memory_model_psot(ns));
+  else if(mm=="irq")
+    memory_model=std::unique_ptr<memory_model_baset>(new memory_model_interruptt(ns));
```
Comment (tautschnig) [inline src/cbmc/bmc.cpp]: Please wrap this line, it's >80 chars long.

Comment (lihaol) [inline src/cbmc/bmc.cpp]: Many thanks for the comments! Done.

**Comment by lihaol** [comment]
I have done a lint cleanup and squashed the commits.

**Comment by tautschnig** [comment]
This needs a rebase.

**Comment by thomasspriggs** [comment]
I am going to close this PR as it appears that work on it has been inactive for an extended period of time. This is not any judgement on how worthwhile this work is. This is part of an effort to reduce the number of open PRs which are not being actively worked on. If you still think it would be worthwhile to get this merged then the branch should be re-based on the latest version of develop and the PR re-opened. Any remaining CI failures will also need to be resolved.

**Comment by martin-cs** [comment]
It is a bit sad to see this one not get merged as it did add a new, useful feature.  I don't know if any of the people doing partial order concurrency (@cesaro @marcelosousa @cristina-david or @tautschnig ) would have some bandwidth to pick this one back up.  It might not be too much work to get merged.

**Pull request closed without merging**